runtime/runtime.go

```diff
@@ -52,75 +48,97 @@ func init() {
 	}
 }
 
+func parseFailpoints(fps string) (map[string]string, error) {
+	// The format is <FAILPOINT>=<TERMS>[;<FAILPOINT>=<TERMS>]*
+	fpMap := map[string]string{}
+	for _, fp := range strings.Split(fps, ";") {
+		fpTerm := strings.Split(fp, "=")
+		if len(fpTerm) != 2 {
+			err := fmt.Errorf("bad failpoint %q", fp)
+			return nil, err
+		}
+		fpMap[fpTerm[0]] = fpTerm[1]
+	}
+	return fpMap, nil
+}
+
 // Enable sets a failpoint to a given failpoint description.
 func Enable(failpath, inTerms string) error {
-	unlock, err := enableAndLock(failpath, inTerms)
-	if unlock != nil {
-		unlock()
-	}
-	return err
+	failpointsMu.Lock()
+	defer failpointsMu.Unlock()
+	return enable(failpath, inTerms)
 }
 
-// enableAndLock enables a failpoint and returns a function to unlock it
-func enableAndLock(failpath, inTerms string) (func(), error) {
+// enable enables a failpoint
+func enable(failpath, inTerms string) error {
+	fp := failpoints[failpath]
+	if fp == nil {
+		return ErrNoExist
+	}
+
 	t, err := newTerms(failpath, inTerms)
 	if err != nil {
 		fmt.Printf("failed to enable \"%s=%s\" (%v)\n", failpath, inTerms, err)
-		return nil, err
-	}
-	failpointsMu.RLock()
-	fp := failpoints[failpath]
-	failpointsMu.RUnlock()
-	if fp == nil {
-		return nil, ErrNoExist
+		return err
 	}
-	fp.mu.Lock()
 	fp.t = t
-	return func() { fp.mu.Unlock() }, nil
+
+	return nil
 }
 
 // Disable stops a failpoint from firing.
 func Disable(failpath string) error {
-	failpointsMu.RLock()
+	failpointsMu.Lock()
+	defer failpointsMu.Unlock()
+	return disable(failpath)
+}
+
+func disable(failpath string) error {
 	fp := failpoints[failpath]
-	failpointsMu.RUnlock()
 	if fp == nil {
 		return ErrNoExist
 	}
 
-	fp.mu.Lock()
-	defer fp.mu.Unlock()
 	if fp.t == nil {
 		return ErrDisabled
 	}
 	fp.t = nil
+
 	return nil
 }
 
 // Status gives the current setting for the failpoint
 func Status(failpath string) (string, error) {
-	failpointsMu.RLock()
+	failpointsMu.Lock()
+	defer failpointsMu.Unlock()
```

Comment:
```suggestion
	failpointsMu.RLock()
	defer failpointsMu.RUnlock()
```

Reply:
This isn't correct. We also need to acquire write lock. Otherwise the server(runtime) might still crash before sending response back to client.